content/post/2020-07-20-keep.source/index.Rmd

```diff
@@ -0,0 +1,124 @@
+This post is for you!
+
+## Where do installed packages live?
+
+Packages are installed 
+
+* at the path you give as `lib` argument to `install.packages()` (or to any `remotes::install_` function, that will pass them on to `install.packages()`);
+
+* most often since you won't give any, at the first path returned by `.libPaths()` that exists and for which the user has the right permissions. There are [several ways to change the paths returned, should you want to do so](https://stackoverflow.com/a/31707983/5489251).[^libraries]
+
+Now at library loading, the important argument is called `lib.loc`, not `lib`. 
+
+`r hugodown::embed_tweet("1275366793423523842")`
+
+Now how do you know where any of your installed packages was installed?
+You can use `find.package()` and `path.package()`!
+
+To check whether a package is installed, it is better to use `find.package()` than `installed.packages()` because the latter, as its docs state, can be slow on some systems.
+In both cases, it does not mean the package is usable, for that you'd need to use `require()`.
+
+## What files are stored locally?
+
+The [R packages book by Hadley Wickham and Jenny Bryan](https://r-pkgs.org/) has [a very neat package called "Package structure and state", including an explanation of the binary state](https://r-pkgs.org/package-structure-state.html#binary-package).
+It says _"There are no .R files in the R/ directory - instead there are three files that store the parsed functions in an efficient file format. This is basically the result of loading all the R code and then saving the functions with save(). (In the process, this adds a little extra metadata to make things as fast as possible)."_
+
+The installed packages in the library are a decompressed version of the binary state.
+There are not all the R files, see [ggplot2 source code](https://github.com/tidyverse/ggplot2/) and ggplot2 on my disk
+
+```{r}
+fs::dir_tree(
+  file.path(
+    find.package("ggplot2"),
+    "R"
+    )
+  )
+```
+
+Under the R folder, there are three files that don't even have the dot R extension!
+
+## How is code stored?
+
+Now, regarding the code, let's mention two important things happening to it.
+
+### Byte compilation
+
+Since [R 3.5](https://cran.r-project.org/doc/manuals/r-release/NEWS.3.html), the code is byte-compiled by default which means it is also stored in a format easier for a machine to deal with.
+You can learn more about byte compilation in the [Efficient R Programming book by Colin Gillespie and Robin Lovelace](https://bookdown.org/csgillespie/efficientR/programming.html#the-byte-compiler), and in [a talk by R Core Member Tomas Kalibera](https://blog.revolutionanalytics.com/2017/08/take-advantage-compiler.html).
+
+### Original formatting and comments?
+
+Then, by default, the source code is stripped of all empty lines and comments because they are useless for code execution and take up space.[^thx]
```

Comment:
Are you sure that this is the order? I.e. first byte compile, then strip?

Reply:
I think that for this to understand, one needs to know that parsed code and function (etc.) objects may have attached information about the source code.